src/serac/physics/solid_mechanics.hpp

```diff
@@ -836,29 +838,102 @@ class SolidMechanics<order, dim, Parameters<parameter_space...>, std::integer_se
+            p_blk.MakeRef(const_cast<mfem::Vector&>(u), disp_size, contact_.numPressureTrueDofs());
+
+            mfem::Vector r_blk, g_blk;
+            r_blk.MakeRef(r, 0, disp_size);
+            g_blk.MakeRef(r, disp_size, contact_.numPressureTrueDofs());
+
+            double dt = 1.0;
+            contact_.setDisplacements(u_blk);
+            // we need to call update first to update gaps
+            contact_.update(1, 1.0, dt);
+            // with updated gaps, we can update pressure for contact pairs with penalty enforcement
+            contact_.setPressures(p_blk);
+            // call update again with the right pressures
+            contact_.update(1, 1.0, dt);
+
+            mfem::Vector res =
+                (*residual_)(u_blk, zero_, shape_displacement_, *parameters_[parameter_indices].state...);
+
+            res.Add(1.0, contact_.trueContactForces());
+
+            r_blk = res;
+            r_blk.SetSubVector(bcs_.allEssentialTrueDofs(), 0.0);
+
+            g_blk.Set(1.0, contact_.trueGaps());
+          },
+
+          // gradient of residual function with contact Jacobian
+          [this](const mfem::Vector& u) -> mfem::Operator& {
+            mfem::Vector u_blk;
+            u_blk.MakeRef(const_cast<mfem::Vector&>(u), 0, displacement_.space().GetTrueVSize());
+            auto [r, drdu] = (*residual_)(differentiate_wrt(u_blk), zero_, shape_displacement_,
+                                          *parameters_[parameter_indices].state...);
+            J_             = assemble(drdu);
+
+            J_contact_ = contact_.contactJacobian();
+            if (J_contact_->IsZeroBlock(0, 0)) {
+              J_contact_->SetBlock(0, 0, J_.release());
+            } else {
+              J_contact_->SetBlock(
+                  0, 0, mfem::Add(1.0, *J_, 1.0, static_cast<mfem::HypreParMatrix&>(J_contact_->GetBlock(0, 0))));
+            }
+
+            J_e_ = bcs_.eliminateAllEssentialDofsFromMatrix(
+                static_cast<mfem::HypreParMatrix&>(J_contact_->GetBlock(0, 0)));
 
-        // gradient of residual function
-        [this](const mfem::Vector& u) -> mfem::Operator& {
-          auto [r, drdu] =
-              (*residual_)(differentiate_wrt(u), zero_, shape_displacement_, *parameters_[parameter_indices].state...);
-          J_   = assemble(drdu);
-          J_e_ = bcs_.eliminateAllEssentialDofsFromMatrix(*J_);
-          return *J_;
-        });
+            return *J_contact_;
+          });
+    }
+  }
+
+  /**
+   * @brief Add a mortar contact boundary condition
+   *
+   * @param pair_id Unique identifier for the ContactPair
+   * @param bdry_attr_surf1 MFEM boundary attributes for the first surface
+   * @param bdry_attr_surf2 MFEM boundary attributes for the second surface
+   * @param contact_opts Defines contact method, enforcement, type, and penalty
+   */
+  void addContactPair(int pair_id, const std::set<int>& bdry_attr_surf1, const std::set<int>& bdry_attr_surf2,
```

Comment:
I would add a comment that this is only implemented for quasi-static simulations. A SLIC_ERROR checking for this would also be a good idea.